From 6a74d1973d0aad3e75efd86f4298318f40f9a728 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:06:46 -0400
Subject: [powerpc] tm: Remove interrupt disable in __switch_to()

Message-id: <1410545655-205645-178-git-send-email-dzickus@redhat.com>
Patchwork-id: 94178
O-Subject: [RHEL7 PATCH 177/626] powerpc/tm: Remove interrupt disable in __switch_to()
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 7ba5fef7d9e1880635cbb2fd698e8a24dc366d0f
Author: Michael Neuling <mikey@neuling.org>
Date:   Wed Oct 2 17:15:14 2013 +1000

    powerpc/tm: Remove interrupt disable in __switch_to()

    We currently turn IRQs off in __switch_to(0 but this is unnecessary as it's
    already disabled in the caller.

    This removes the IRQ disable but adds a check to make sure it is really off
    in case this changes in future.

    Signed-off-by: Michael Neuling <mikey@neuling.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/process.c b/arch/powerpc/kernel/process.c
index 19e7c4c..7734daf 100644
--- a/arch/powerpc/kernel/process.c
+++ b/arch/powerpc/kernel/process.c
@@ -741,12 +741,13 @@ struct task_struct *__switch_to(struct task_struct *prev,
  struct task_struct *new)
 {
  struct thread_struct *new_thread, *old_thread;
- unsigned long flags;
  struct task_struct *last;
 #ifdef CONFIG_PPC_BOOK3S_64
  struct ppc64_tlb_batch *batch;
 #endif
 
+ WARN_ON(!irqs_disabled());
+
  /* Back up the TAR across context switches.
   * Note that the TAR is not available for use in the kernel.  (To
   * provide this, the TAR should be backed up/restored on exception
@@ -866,8 +867,6 @@ struct task_struct *__switch_to(struct task_struct *prev,
  }
 #endif /* CONFIG_PPC_BOOK3S_64 */
 
- local_irq_save(flags);
-
  /*
   * We can't take a PMU exception inside _switch() since there is a
   * window where the kernel stack SLB and the kernel stack are out
@@ -887,8 +886,6 @@ struct task_struct *__switch_to(struct task_struct *prev,
  }
 #endif /* CONFIG_PPC_BOOK3S_64 */
 
- local_irq_restore(flags);
-
  return last;
 }
 
-- 
1.7.1